models/base.py

```diff
@@ -978,6 +981,152 @@ def calculate_coordinates(self):
+        # arguments, not on the array argument.
+
+        # It could probably be made faster by making a first pass doing:
+        #   greatest( ra** ) >= ra AND least( ra** ) <= ra AND
+        #   greatest( dec** ) >= dec AND least( dec** ) <= dec
+        # with indexes in ra** and dec**.  Put the results of that into
+        # a temp table, and then do the polygon search on that temp table.
+        #
+        # I have no clue how to implement that simply here as as an
+        # SQLAlchemy filter, so I implement it in find_containing()
+
+        # return sa.and_( func.greatest( self.ra00, self.ra01, self.ra10, self.ra11 ) >= ra,
+        #                 func.least( self.ra00, self.ra01, self.ra10, self.ra11 ) <= ra,
+        #                 func.greatest( self.dec00, self.dec01, self.dec10, self.dec11 ) >= dec,
+        #                 func.least( self.dec00, self.dec01, self.dec10, self.dec11 ) <= dec )
+        return func.q3c_poly_query( ra, dec, sqlarray( [ self.ra00, self.dec00, self.ra01, self.dec01,
+                                                         self.ra11, self.dec11, self.ra10, self.dec10 ] ) )
+
+    @classmethod
+    def find_containing( cls, siobj, session=None ):
+        """Return all images (or whatever) that contain the given SpatiallyIndexed thing
+
+        Parameters
+        ----------
+          siobj: SpatiallyIndexed
+            A single object that is spatially indexed
+
+        Returns
+        -------
+           An sql query result thingy.
+
+        """
+
+        # Overabundance of caution to avoid SQL injection
+        ra = float( siobj.ra )
+        dec = float( siobj.dec )
+
+        with SmartSession( session ) as sess:
+            sess.execute( sa.text( f"SELECT i.id, i.ra00, i.ra01, i.ra10, i.ra11, "
+                                   f"       i.dec00, i.dec01, i.dec10, i.dec11 "
+                                   f"INTO TEMP TABLE temp_find_containing "
+                                   f"FROM {cls.__tablename__} i "
+                                   f"WHERE GREATEST(i.ra00, i.ra01, i.ra10, i.ra11 ) >= :ra "
+                                   f"  AND LEAST(i.ra00, i.ra01, i.ra10, i.ra11 ) <= :ra "
+                                   f"  AND GREATEST(i.dec00, i.dec01, i.dec10, i.dec11 ) >= :dec "
+                                   f"  AND LEAST(i.dec00, i.dec01, i.dec10, i.dec11 ) <= :dec" ),
+                          { 'ra': ra, 'dec': dec } )
+            query = sa.text( f"SELECT i.id FROM temp_find_containing i "
+                             f"WHERE q3c_poly_query( {ra}, {dec}, ARRAY[ i.ra00, i.dec00, i.ra01, i.dec01, "
+                             f"                                          i.ra11, i.dec11, i.ra10, i.dec10 ])" )
+            objs = sess.scalars( sa.select( cls ).from_statement( query ) ).all()
+            sess.execute( sa.text( "DROP TABLE temp_find_containing" ) )
+            return objs
```

Comment:
I don't think it's worth the time figuring out the arcane sqlalchemy of the right subquery.  SQLAlchemy is enough of a nightmare trying to figure out how to do things that are straightfoward in SQL already.  Also, we won't *really* know which we want until we can test it on a big enough database to see the speed trials.

I'm sure we can get something that produces the same results, but the point of using the temporary table was so that the query that didn't have indexes to take advantage of would be working on a much smaller table than the full images table.

Reply:
sounds good to me.